src/Route.php

```diff
@@ -0,0 +1,419 @@
+				$vars[]  = substr($segment, 1);
+				$regex[] = '(.*)';
+			}
+			elseif (isset($segment[0]) && $segment[0] == '\\' && $segment[1] == '*')
+			{
+				// Match an escaped splat segment.
+				$regex[] = '\*' . preg_quote(substr($segment, 2));
+			}
+			elseif ($segment == ':')
+			{
+				// Match an unnamed variable without capture.
+				$regex[] = '([^/]*)';
+			}
+			elseif (isset($segment[0]) && $segment[0] == ':')
+			{
+				// Match a named variable and capture the data.
+				$varName = substr($segment, 1);
+				$vars[]  = $varName;
+
+				// Use the regex in the rules array if it has been defined.
+				$regex[] = array_key_exists($varName, $this->getRules()) ? '(' . $this->getRules()[$varName] . ')' : '([^/]*)';
+			}
+			elseif (isset($segment[0]) && $segment[0] == '\\' && $segment[1] == ':')
+			{
+				// Match a segment with an escaped variable character prefix.
+				$regex[] = preg_quote(substr($segment, 1));
+			}
+			else
+			{
+				// Match the standard segment.
+				$regex[] = preg_quote($segment);
+			}
+		}
+
+		$this->setRegex(chr(1) . '^' . implode('/', $regex) . '$' . chr(1));
+		$this->setRouteVariables($vars);
+	}
+
+	/**
+	 * Retrieve the controller which handles this route
+	 *
+	 * @return  mixed
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public function getController()
+	{
+		return $this->controller;
+	}
+
+	/**
+	 * Retrieve the default variables defined by the route
+	 *
+	 * @return  array
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public function getDefaults(): array
+	{
+		return $this->defaults;
+	}
+
+	/**
+	 * Retrieve the HTTP methods this route supports
+	 *
+	 * @return  string[]
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public function getMethods(): array
+	{
+		return $this->methods;
+	}
+
+	/**
+	 * Retrieve the route pattern to use for matching
+	 *
+	 * @return  string
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public function getPattern(): string
+	{
+		return $this->pattern;
+	}
+
+	/**
+	 * Retrieve the path regex this route processes
+	 *
+	 * @return  string
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public function getRegex(): string
+	{
+		if (!$this->regex)
+		{
+			$this->buildRegexAndVarList();
+		}
+
+		return $this->regex;
+	}
+
+	/**
+	 * Retrieve the variables defined by the route
+	 *
+	 * @return  array
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public function getRouteVariables(): array
+	{
+		if (!$this->regex)
+		{
+			$this->buildRegexAndVarList();
+		}
+
+		return $this->routeVariables;
+	}
+
+	/**
+	 * Retrieve the regex rules keyed using the route variables
+	 *
+	 * @return  array
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public function getRules(): array
+	{
+		return $this->rules;
+	}
+
+	/**
+	 * Set the controller which handles this route
+	 *
+	 * @param   mixed  $controller  The controller which handles this route
+	 *
+	 * @return  $this
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public function setController($controller): self
```

Comment:
Can we typehint this with callable? Or do we want to allow setting a class here?

Reply:
It removes some capability from https://github.com/joomla-framework/application/pull/81 if we restrict to callable.  I don't think the router should enforce types on controllers personally, leave that to something closer to the application.